man/man3/allocf.3
.\" 
.\" Copyright (c) 2008 Gordon D. Carrie.  All rights reserved.
.\" 
.\" Licensed under the Open Software License version 3.0
.\" 
.\" Please address questions and feedback to user0@tkgeomap.org
.\" 
.\" $Id: allocf.3,v 1.5 2008/11/25 23:03:21 gcarrie Exp $
.\"
.TH allocf 3 "Dynamic multidimensional allocators"
.SH NAME
MALLOC, CALLOC, REALLOC, FREE \- allocators with optional diagnostics
mallocf2, freef2, mallocf3, freef3, mallocf4, freef4 \- dynamic allocators for multidimensional arrays of floats
.SH SYNOPSIS
.nf
\fB#include "allocf.h"\fP
\fBfloat **mallocf2(size_t\fP \fIj\fP, \fBsize_t\fP \fIi\fP);
\fBvoid freef2(float **\fP\fId\fP);
\fBfloat ***mallocf3(size_t\fP \fIk\fP, \fBsize_t\fP \fIj\fP, \fBsize_t\fP \fIi\fP);
\fBvoid freef3(float ***\fP\fId\fP);
\fBfloat ****mallocf4(size_t\fP \fIl\fP, \fBsize_t\fP \fIk\fP, \fBsize_t\fP \fIj\fP, \fBsize_t\fP \fIi\fP);
\fBvoid freef4(float ****\fP\fId\fP);
.fi
.SH DESCRIPTION
These allocators create multidimensional arrays of float values.
.PP
\fBmallocf2\fP allocates a two dimensional array of float values.  If the
return value is assigned to \fId\fP, then the highest index of the array
will be \fId\fP[\fIj\fP-1][\fIi\fP-1].
\fId\fP[\fB0\fP] will be a contiguous block of \fIj\fP \fB*\fP \fIi\fP floats.
If something goes wrong, the function generates an error message that can be
retrieved with a call to \fBerr_get\fP, and returns \fBNULL\fP.
\fBfreef2\fP frees all memory associated with \fId\fP,
which should be a return value of \fBmalloc2f\fP.
.PP
\fBmallocf3\fP allocates a three dimensional array of float values.  If the
return value is assigned to \fId\fP, then the highest index of the array
will be \fId\fP[\fIk\fP-1][\fIj\fP-1][\fIi\fP-1].
\fId\fP[\fB0\fP][\fB0\fP] will be a contiguous block of
\fIk\fP \fB*\fP \fIj\fP \fB*\fP \fIi\fP floats.
If something goes wrong, the function generates an error message that can be
retrieved with a call to \fBerr_get\fR, and returns \fBNULL\fR.
\fBfreef3\fP frees all memory associated with \fId\fP,
which should be a return value of \fBmalloc3f\fP.
.PP
\fBmallocf4\fP allocates a four dimensional array of float values.  If the
return value is assigned to \fId\fP, then the highest index of the array
will be \fId\fP[\fIl\fP-1][\fIk\fP-1][\fIj\fP-1][\fIi\fP-1].
terminated.
\fId\fP[\fB0\fP][\fB0\fP][\fB0\fP]  will be a contiguous block of
\fIl\fP \fB*\fP \fIk\fP \fB*\fP \fIj\fP \fB*\fP \fIi\fP floats.
If something goes wrong, the function generates an error message that can be
retrieved with a call to \fBerr_get\fR, and returns \fBNULL\fR.
\fBfreef4\fP frees all memory associated with \fId\fP,
which should be a return value of \fBmalloc4f\fP.
.SH SEE ALSO
\fBMALLOC\fP (3), \fBCALLOC\fP (3), \fBREALLOC\fP (3), \fBFREE\fP (3),
\fBerr_msg\fP (3)
.SH AUTHOR
Gordon Carrie (user0@tkgeomap.org)


src/allocf.c
/*
 * allocf.c --
 *
 *	This file defines functions that allocate multi-dimensional
 *	arrays of floating point values.  See allocf (3).
 * 
 * Copyright (c) 2008 Gordon D. Carrie
 *
 * Licensed under the Open Software License version 3.0
 *
 * Please send feedback to user0@tkgeomap.org
 *
 * $Id: allocf.c,v 1.14 2008/11/19 04:03:23 gcarrie Exp $
 *
 **********************************************************************
 *
 */

#include "alloc.h"
#include "err_msg.h"
#include "allocf.h"

float ** mallocf2(long j, long i)
{
    float **dat = NULL;
    long n;
    size_t jj, ii;		/* Addends for pointer arithmetic */
    size_t ji;

    if (j <= 0 || i <= 0) {
	err_append("Array dimensions must be positive.\n");
	return NULL;
    }
    jj = (size_t)j;
    ii = (size_t)i;
    ji = jj * ii;
    if (ji / jj != ii) {
	err_append("Dimensions too big for pointer arithmetic.\n");
	return NULL;
    }
    dat = (float **)CALLOC(jj + 2, sizeof(float *));
    if ( !dat ) {
	err_append("Could not allocate memory.\n");
	return NULL;
    }
    dat[0] = (float *)CALLOC(ji + 1, sizeof(float));
    if ( !dat[0] ) {
	FREE(dat);
	err_append("Could not allocate memory.\n");
	return NULL;
    }
    for (n = 1; n <= j; n++) {
	dat[n] = dat[n - 1] + i;
    }
    return dat;
}

void freef2(float **dat)
{
    if (dat && dat[0]) {
	FREE(dat[0]);
    }
    FREE(dat);
}

float *** mallocf3(long k, long j, long i)
{
    float ***dat = NULL;
    long n, ne;
    size_t kk, jj, ii;		/* Addends for pointer arithmetic */
    size_t kj, kji;


    if (k <= 0 || j <= 0 || i <= 0) {
	err_append("Array dimensions must be positive.\n");
	return NULL;
    }
    kk = (size_t)k;
    jj = (size_t)j;
    ii = (size_t)i;
    kj = kk * jj;
    kji = kj * ii;
    if (kj / kk != jj || kji / kj != ii) {
	err_append("Dimensions too big for pointer arithmetic.\n");
	return NULL;
    }
    dat = (float ***)CALLOC(kk + 1, sizeof(float **));
    if ( !dat ) {
	err_append("Could not allocate memory.\n");
	return NULL;
    }
    dat[0] = (float **)CALLOC(kj + 1, sizeof(float *));
    if ( !dat[0] ) {
	FREE(dat);
	err_append("Could not allocate memory.\n");
	return NULL;
    }
    dat[0][0] = (float *)CALLOC(kji + 1, sizeof(float));
    if ( !dat[0][0] ) {
	FREE(dat[0]);
	FREE(dat);
	err_append("Could not allocate memory.\n");
	return NULL;
    }
    for (n = 1; n < k; n++) {
	dat[n] = dat[n - 1] + j;
    }
    for (n = 1, ne = k * j; n < ne; n++) {
	dat[0][n] = dat[0][n - 1] + i;
    }
    return dat;
}

void freef3(float ***dat)
{
    if (dat) {
	if (dat[0]) {
	    if (dat[0][0]) {
		FREE(dat[0][0]);
	    }
	    FREE(dat[0]);
	}
	FREE(dat);
    }
}

float **** mallocf4(long l, long k, long j, long i)
{
    float ****dat;
    long n, ne;
    size_t ll, kk, jj, ii;	/* Addends for pointer arithmetic */
    size_t lk, lkj, lkji;

    if (l <= 0 || k <= 0 || j <= 0 || i <= 0) {
	err_append("Array dimensions must be positive.\n");
	return NULL;
    }
    ll = (size_t)l;
    kk = (size_t)k;
    jj = (size_t)j;
    ii = (size_t)i;
    lk = ll * kk;
    lkj = lk * jj;
    lkji = lkj * ii;
    if (lk / ll != kk || lkj / lk != jj || lkji / lkj != ii) {
	err_append("Dimensions too big for pointer arithmetic.\n");
	return NULL;
    }
    dat = (float ****)CALLOC(ll + 1, sizeof(float ***));
    if ( !dat ) {
	err_append("Could not allocate memory.\n");
	return NULL;
    }
    dat[0] = (float ***)CALLOC(lk + 1, sizeof(float **));
    if ( !dat[0] ) {
	FREE(dat);
	err_append("Could not allocate memory.\n");
	return NULL;
    }
    dat[0][0] = (float **)CALLOC(lkj + 1, sizeof(float *));
    if ( !dat[0][0] ) {
	FREE(dat[0]);
	FREE(dat);
	err_append("Could not allocate memory.\n");
	return NULL;
    }
    dat[0][0][0] = (float *)CALLOC(lkji + 1, sizeof(float));
    if ( !dat[0][0][0] ) {
	FREE(dat[0][0]);
	FREE(dat[0]);
	FREE(dat);
	err_append("Could not allocate memory.\n");
	return NULL;
    }
    for (n = 1; n < l; n++) {
	dat[n] = dat[n - 1] + k;
    }
    for (n = 1, ne = l * k; n < ne; n++) {
	dat[0][n] = dat[0][n - 1] + j;
    }
    for (n = 1, ne = l * k * j; n < ne; n++) {
	dat[0][0][n] = dat[0][0][n - 1] + i;
    }
    return dat;
}

void freef4(float ****dat)
{
    if (dat) {
	if (dat[0]) {
	    if (dat[0][0]) {
		if (dat[0][0][0]) {
		    FREE(dat[0][0][0]);
		}
		FREE(dat[0][0]);
	    }
	    FREE(dat[0]);
	}
	FREE(dat);
    }
}
